### ISSUE TITLE
New maintainer for this project

### ISSUE THREAD

**Issue by StefanKert** [StefanKert/BuildVision#21]
Hey guys,

like I´ve been mentioning in #17 I´m planning to becoming the maintainer for this project. A colleague of mine @tschmiedlechner and me have been working through the code for the last four weeks and now we think that we should take the next step and make it official.

If @nagits agrees we would like to take over this project and try to bring in the ideas and thoughts we gathered in the last weeks and also trying to fix the Issues that are opened till today.

## Current State

We´ve started to migrate the Project do VS 2017. We released a version 1.5.0 that works with VS 2017. You can grab it under:

https://github.com/StefanKert/BuildVision/releases/tag/v1.5.0

@nagits was working on fixing #5 on another branch. We´ve tested these changes and merged them into master. We are planning to make a release with this issue solved until the end of this week.

### Restructuring the codebase
Furthermore we are now working on refactoring and re-engineering the project to make it more testable and to make development easier for us and for the community. For this purpose we are planning to achieve the following goals:

- [x] Restructure code base to use multiple projects
- [x] Make usercontrols and views testable without running Visual Studio (separate Testproject)
- [x] Migrating to [Microsoft.Build.Evaluation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.evaluation) for easier Project property resolution (will drop support for VS 2012)

### Bugfixes and new features

Like i mentioned before we´ve also collected some ideas during going through the sources and through the open issues. This list can be extended, but for the sake of shortness I´m only mentioning a few:

- [x] **Bugfix**: Build finish time is not correct - See #18 for details
- [x ] **Bugfix**:  Long multi-line error messages get truncated -See #13 for details (Ideas for UI integration are appreciated)
- [x] **Bugfix**: Sort doesn't update during build - See #5 for details 
- **Performance**: I guess this is a pretty big blocker for many devs to use BuildVision. (#7, #8). For this purpose we need to get more Information if and when BuildVision is slowing down the build for you. 
- **New Feature**: Add new building state: deploying - See #14 for details
- **New Feature**: Optionally hide 'UpToDate' targets - See #15  for details
- [x] **New Feature**: Adding support for new Csproj libraries
- **New Feature**: Toggle Buttons to display succeeded Projects or failed Projects only.

## Next Steps

So as you can imagine the next steps will be to take over this repo. I hope this is ok for @nagits and maybe we can get in touch in the near future.

## Contact

I create this issue to start this discussion on the future of this product. If you have any specific questions you can find me on Twitter [@StefanKert](https://twitter.com/StefanKert) or send me an email [stefan.kert@sbg.at](stefan.kert@sbg.at).  

## Enclosing

I´d like to thank @nagits for the work he put in this awesome project used by thousands of people. I really like the way it enriches my workflow and the build experience in Visual Studio. This is also one of the many reasons I want to keep up the good work here so that we can improve this project together. Furthermore I want to thank all of you for the future discussions, contributions and hints. 

# tl;dr
I´m planning to take over responsibility for BuildVision.

**Comment by nagits** [comment]
Oh, guys...

I'm really happy that my work are not wasted and the project is used by quite a lot of people.

I stopped supporting this project, because I no longer use VisualStudio & C# in my work and I don't have VisualStudio installed on my computer.

Yes, of course, I'm not against any initiative which will save life of this tool - I will be doubly happy!
@StefanKert, @tschmiedlechner, Great job :+1: :+1:!,
Please tell me how I can help you to take BuildVision under your wing.

p.s. Sorry @StefanKert, I didn't get your letter (https://github.com/nagits/BuildVision/issues/17#issuecomment-288970902).

**Comment by StefanKert** [comment]
Hey @nagits,

thanks for responding that fast 👍. Nevermind maybe sent it to the wrong email address. I guess it would be helpful to work out a migration plan that includes all the pieces that need to be moved.

- [ ] GitHub Repository
- [ ] Appveyor (I got a running build already so probably we can take that one [Build](https://ci.appveyor.com/project/StefanKert/buildvision))
- [ ] [Visual Studio Marketplace](https://marketplace.visualstudio.com/items?itemName=AlekseyNagovitsyn.BuildVision) 

These are the things that come in my mind right now. Maybe there are further things that need to be moved? 

Anyway I really appreciate the work you´ve done and with this project and hope that we can keep up the good work!

**Comment by NickCraver** [comment]
Very happy to see this happen - we're already running on VS 2017 internally thanks to https://github.com/Konamiman/BuildVision (@Konamiman will be excited this is getting love). Let us know if we can help.

**Comment by StefanKert** [comment]
Thanks @NickCraver for the offer. I am really glad that we are already getting positive feedback on our plans. What I think is really needed is some more feedback on the work ideas, or problems people have with BuildVision. We will try to get the community in for this project and enable devs to contribute here more easily

Like I´ve been mentioning we are planning to get to a point where getting started with the source of  BuildVision should be very easy for everyone. We have already started a fork where we are refactoring and restructuring the code base  to make the structure more obvious and easier to understand. 

Another main goal of this restructuring is to get a separate project for all WPF Components. This then should enable us to make a Test project that instantiates this WPF Components with dummy data so that we can make testing possible and pretty easy for everyone.

Therefore we will have a test (or another exe) that spawns two windows. A window that includes the BuildVision Usercontrol and another Window that consists of buttons that trigger the needed events (Solution loaded, Starting Build, Build failed etc.). The positive effect should be a faster dev time, because there´s no need to start Visual Studio each time we are debugging. Also it should be easier to contribute to this project, because the Views are separated and even if you have no knowledge of visual studio extension development (VSED) you should be able to contribute to this project. 

Another thing we will get rid of are the many references do EnvDTE and other VSED components that are tightly coupled to the ViewModels and many other components in the system. Therefore we will create a single wrapper that should return a Model that will then be passed through the program. To make this more obvious there should be a separate Project that is used for getting all the data that is needed and also for binding events.

In the end this should result in the following project structure:

- BuildVision: The main project. This will include all componentas that are needed to make the project work as a Visual Studio Extension. Configuration files, the class that inherits from Package. This project will need some specific knowledge of VSED.

- BuildVision.Core: This project will include shared components like logging, ViewModelBase etc. There should be no references to an VSED and also there is no specific knowledge for VSED needed.

- BuildVision.View: All view elements should be included in this project. Again no specific knowledge for VSED is needed here. If you know WPF, if you want to try out something with WPF or if you are an expert with WPF you should be good to go to contribute here. We are still discussing if we should keep ViewModels also inside this project. 

- BuildVision.Contracts: Here we keep all the contracts that are used within BuildVision. Exceptions, POCOs and those things should be kept here. No specific knowledge needed. Only Plain C# 👍 

- BuildVision.VisualStudioConnection: This project will be the layer that connects our Project to the visual studio extension system. Here we have references to VSED components. We will try to keep it clear but specific VSED knowledge will be needed here.

To summarize the our main goal:

# We are planning to make contributions to BuildVision easy. 
For this purpose we will try to find Issues that can be labeld either with "Firsttimersonly" or "Upforgrabs". We really want to get the community in and contribute to this project.

**Comment by NickCraver** [comment]
@StefanKert That sounds like a good split of concerns to me. As a user, my main issue with it today is the performance impact. I think that'll be insanely easier to keep an eye on and have tests for with this split. Please ping when you're closer on this - I'd love to dig into some things there and this layout would save a ton of time.

**Comment by StefanKert** [comment]
@NickCraver Yep performance is a big impact right now. We´ve already spotted some parts in the source code that might cause BuildVision to slow down. Will get back to you as soon as we´ve some results.

**Comment by StefanKert** [comment]
Hey guys,

just wanted to give you a quick update on the current state of our "migration".

We´ve refactored the codebase so that the proposed project architecture is implemented now. It is possible to start the Extensionview as seperate Window now in a seperate Testproject, so no need for launching another Visual Studio instance. For now it is not possible to add a solution or a project to the standalone window, but this should be implemented pretty soon. It is also possible now to launch the settings windows as standalone windows so that we can improve the quality of those more easily.

There is still an issue with the standalone window that we didn´t come up with a solution for now. Since this is launched outside of VS you cannot get the colors that are used in VS so the Window is now using all the default Systemcolors. Maybe someone has an idea here.

If anyone wants to throw a glimpse on the new project structure you can find it here:

[https://github.com/StefanKert/BuildVision/tree/dev](https://github.com/StefanKert/BuildVision/tree/dev)

I will not make any bigger changes in the next three weeks since I´ve got a lot on my plate but then we will start to do further minor refactorings. After these changes we should be able to send a Pull Request and merge the new repo into the old @nagits.

**Comment by StefanKert** [comment]
@nagits Since we´ve got the Pull Request merged successfully I wanted to get back to the steps we should take to get this up and running again.

As I mentioned before we need to proceed with the following steps:

- [ ] GitHub Repository 
- [ ] Appveyor (I got a running build already so probably we can take that one [Build](https://ci.appveyor.com/project/StefanKert/buildvision))
- [ ] [Visual Studio Marketplace](https://marketplace.visualstudio.com/items?itemName=AlekseyNagovitsyn.BuildVision) 

I don´t know if you want to stay the owner for this repo or if it is more convenient to transfer ownership over to me.  For the Visual Studio Marketplace extension and the Appveyor we also need to make a decision on who is responsible for the maintenance. 

I don´t know what you are planning for the future and how much effort you´d like to put in this project, or how involved you´d like to be.

**Comment by nagits** [comment]
@StefanKert 
1. Let's transfer repo ownership over to you.
2. OK.
3. I've added user stefan.kert@sbg.at as Collaborator into the current publisher account on marketplace.visualstudio.com (visualstudiogallery.msdn.microsoft.com).

**Comment by nagits** [comment]
Error: `StefanKert already has a repository in the nagits/BuildVision network.`
Probably, you need to remove the fork. Renaming is not a solution.

**Comment by StefanKert** [comment]
@nagits great! Sry i removed the repo. The transfer should work now

**Comment by Itori** [comment]
@StefanKert with the deletion of your repo, the tag 1.5.0 doesn't exist anymore.
 Could you create the release back ?

**Comment by StefanKert** [comment]
@Itori  Thanks dude for pointing this out..

You can grab a running version by following this [link](https://github.com/StefanKert/BuildVision/releases/tag/untagged-e32c244fb98652d81e0f )

**Comment by igorushi** [comment]
@StefanKert, cool!

**Comment by dantz15** [comment]
Glad to know that this project is active again guys! Keep it up.

**Comment by StefanKert** [comment]
Since we got v2.0.0 out I´m closing this. Moved all tasks that still need to be done to separate issues. Thanks for all of your help and your feedback. Really appreciate it!! Special thanks to @nag5000 for the support!
